_posts/en/newsletters/2019-03-12-newsletter.md

```diff
@@ -0,0 +1,273 @@
+  created a block with two transactions whose 32-byte hashes (txids),
+  when combined together, appear to be a 64-byte transaction, it's
+  possible to create two different interpretations of the merkle root in
+  the block header---one where it points to a single invalid 64-byte
+  transaction and one where it points to two valid transactions.
+
+    FIXME: image
+
+    This can create a problem for Bitcoin Core as, normally, if it
+    rejects a block as being invalid, it will add the header hash of
+    that block to a cache so that it doesn't waste resources requesting
+    or re-processing that block again.  This allowed an attacker to
+    send your node the invalid form of the block to subsequently prevent
+    your node from processing its valid form or any blocks that descend
+    from it, forking you off the chain.
+
+    A similar vulnerability was disclosed in 2012 as [CVE-2012-2459][]
+    and Bitcoin Core was adapted then to not cache invalidity for blocks
+    that have multiple interpretations.  However, an
+    [optimization][bitcoin core #7225] introduced in Bitcoin Core 0.13.0
+    reintroduced this basic problem and necessitated a [fix][bitcoin
+    core #9765], which was included in Bitcoin Core 0.14.0.  Daftuar's
+    email includes a [very informative PDF][daftuar pdf] that not only
+    describes this specific problem in detail and shows its cost to be a
+    very small 30 bits of brute force work (although you also need to
+    mine a custom block) but which also describes other known
+    vulnerabilities possible with Bitcoin's merkle trees and calculates
+    the average amount of brute force work to exploit them.  Daftuar did
+    not find any instances of the attack to date in the current
+    consensus block chain.
+
+- **Cleanup soft fork proposal discussion:** this week saw discussion
+  about the consensus cleanup soft fork [proposal][bip-cleanup]
+  described in [last week's newsletter][newsletter #36].  Russell
+  O'Connor [raised the concern][roconnor codesep] that the
+  `OP_CODESEPARATOR` rule in the proposed soft fork could prevent
+  existing UTXOs from being spent.  It's not possible to detect this
+  because people could've paid money to a P2SH address whose
+  not-yet-revealed redeemScript uses the discouraged opcode.  O'Connor
+  proposes to mitigate the problem of `OP_CODESEPARATOR` being used to
+  increase worse-case block verification time by instead increasing the
+  weight (vbytes) of transactions whose evaluated scripts contain the
+  opcode.  This would reduce the maximum number of code separators that
+  could be contained within a block while also likely reducing the
+  overall size and total number of operations in the block to the point
+  where it could be verified reasonably quickly.
+
+    O'Connor also raised a [similar concern][roconnor sighash] regarding
+    the soft fork's proposal to remove unallocated sighash type bytes.
```

Comment:
Would it be better to say "treat unallocated sighash type bytes as invalid"? (vs. removing bytes)